### PR TITLE
Bye, bye {V8}; Hello, {Rcpp}!!

### PR DESCRIPTION
**Pull request by hrbrmstr**
I _think_ I dotted all the i's and crossed all the t's.

- Removed all V8 bits in src and tests
- Added in Rcpp dep
- Added in R script to `data-raw` which will fetch the latest darksky stringified tree and munge it into something that can be used in Rcpp and auto-generates `src/generated-vars.h`.
- Modified one test to use `NA_character_` instead of `NA`
- Re-built Rmd
- Bumped up version
- Added myself to DESCRIPTION and changed a teensy bit of text in there and README.Rmd to note the change to Rcpp.
- Ran all tests 
- Ran CRAN checks

### CODE REVIEW

**Comment by hrbrmstr** [comment]
now we'll see how Travis like this mod

**Comment by hrbrmstr** [comment]
Lemme port https://github.com/darkskyapp/tz-lookup/blob/master/test.js over to here and run the full checks. (later this week)

**Comment by hrbrmstr** [comment]
OK, so the `mocha` tests all pass for the darksky JS lib but some of them fail (i did a way too quick port of them, so it cld be my js array data wrangling) for the Rcpp port so i def have to take more time on this but will also run them against the accurate functions as well. shld be a robust test suite!

**Comment by ateucher** [comment]
@hrbrmstr  this is absolutely brilliant, thank you so much! I definitely will take it. Thinking I'll take this now and when/if you get the time for another PR with the mocha tests that would be great. Do you have a start somewhere I can look at?

File: README.Rmd [hunk 1]
```diff
@@ -18,8 +18,7 @@ knitr::opts_chunk$set(
 
 Input latitude and longitude values or an `sf/sfc` POINT object and get back 
 the timezone in which they exist. Two methods are implemented. One is very 
-fast and uses the *V8* package to access the 
-[`tz-lookup.js` javascript library](https://github.com/darkskyapp/tz-lookup/). 
+fast and uses an Rcpp in conjunction with source data from (https://github.com/darkskyapp/tz-lookup/).
```
Comment (ateucher) [inline README.Rmd]:
```suggestion
fast and uses Rcpp in conjunction with source data from (https://github.com/darkskyapp/tz-lookup/). 
```

File: README.md [hunk 2]
```diff
@@ -12,11 +12,10 @@ Status](https://img.shields.io/codecov/c/github/ateucher/lutz/master.svg)](https
 
 Input latitude and longitude values or an `sf/sfc` POINT object and get
 back the timezone in which they exist. Two methods are implemented. One
-is very fast and uses the *V8* package to access the [`tz-lookup.js`
-javascript library](https://github.com/darkskyapp/tz-lookup/). However,
-speed comes at the cost of accuracy - near time zone borders away from
-populated centres there is a chance that it will return the incorrect
-time zone.
+is very fast and uses an Rcpp in conjunction with source data from
```
Comment (ateucher) [inline README.md]:
```suggestion
is very fast and uses Rcpp in conjunction with source data from
```

**Comment by hrbrmstr** [comment]
I am liking that the fam sleeps late on holiday :-)

Running:

```bash
$ mocha test.js
```

in the `tz-lookup` repo runs 1,028 tests designed to catch what look like edge cases.

I made most of them into a CSV file: [tz-cases.csv.gz](https://github.com/ateucher/lutz/files/3325337/tz-cases.csv.gz) and ran both 'fast' and 'accurate' tests. 'fast' has 193 test errors and 'accurate' has 126:

```r
library(lutz)
library(tidyverse)
```

Read in the cases

```r
cases <- read_csv("~/Data/tz-cases.csv.gz", col_types="ddc")
```

Do tests with ‘fast’

```r
fast <- tz_lookup_coords(cases$lat, cases$lng, "fast", FALSE)
fast_err <- which(!(cases$val == fast))
```

Do tests with ‘accurate’

```r
accurate <- tz_lookup_coords(cases$lat, cases$lng, "accurate", FALSE)
accurate_err <- which(!(cases$val == accurate))
```

Gather all the errors

```r
bind_rows(
  cases[fast_err,] %>%
    mutate(case_num = fast_err, method = "fast") %>%
    select(case_num, method, everything()),
  cases[accurate_err,] %>%
    mutate(case_num = accurate_err, method = "accurate") %>%
    select(case_num, method, everything())
) -> tz_errs
```

See how many for each test

```r
count(tz_errs, method)
## # A tibble: 2 x 2
##   method       n
##   <chr>    <int>
## 1 accurate   126
## 2 fast       193
```

See if there are common errors

```r
count(tz_errs, case_num, sort = TRUE)
## # A tibble: 315 x 2
##    case_num     n
##       <int> <int>
##  1      573     2
##  2      596     2
##  3      846     2
##  4      940     2
##  5        4     1
##  6        7     1
##  7        9     1
##  8       12     1
##  9       14     1
## 10       15     1
## # … with 305 more rows
```

I'll poke at the C++ code a bit to see what I'm not translating from javascript properly.

**Comment by hrbrmstr** [comment]
One reason 'accurate' results differ is some return GMT offsets vs what the test functions expect. 

However, some are errors like: `Antarctica/Davis` which shld be +7 unless I'm reading the local timezone db wrong. 

```r
# A tibble: 126 x 6
   case_num method     lat     lng val                err_val   
      <int> <chr>    <dbl>   <dbl> <chr>              <chr>     
 1       77 accurate  90    -90    Etc/GMT            Etc/GMT+6 
 2       79 accurate  90     90    Etc/GMT            Etc/GMT-6 
 3      525 accurate -79.0   81.4  Antarctica/Davis   Etc/GMT-5 
 4      526 accurate -73.1  107.   Antarctica/Vostok  Etc/GMT-7 
 5      528 accurate -73.4    5.44 Antarctica/Troll   Etc/GMT   
 6      529 accurate -66.5   64.1  Antarctica/Mawson  Etc/GMT-4 
 7      534 accurate -81.8   20.3  Antarctica/Troll   Etc/GMT-1 
 8      535 accurate -89.7  -12.8  Antarctica/McMurdo Etc/GMT+1 
 9      539 accurate -66.3  -77.8  Antarctica/Rothera Etc/GMT+5 
10      549 accurate -80.9 -167.   Antarctica/McMurdo Etc/GMT+11
# … with 116 more rows
```

This does bring up another question: do both methods need to return the exact same string for the exact same input vals? GMT offsets are fine timezone specifiers IMO but it'll make testing and comparing a tad more grungy. We may want to note the diff vals in docs somewhere.

**Comment by hrbrmstr** [comment]
Looks like (re: 'fast' errs) might have something to do with the double math ops. I'll take a look at this again on the morrow.

**Comment by hrbrmstr** [comment]
Yep. truncated a double prematurely.

```
✔ | 1020       | tz-lookup js edge cases [0.7 s]
```

Will update the PR with these then we can discuss the other bits abt tz strings being returned by each method.

File: README.md (hunk 2 above)
Comment (hrbrmstr) [reply README.md]: fixed in forthcoming PR update

**Comment by hrbrmstr** [comment]
I rly cld not stand the line lengths for the generated data so i added a call to `fold` (linux/macOS) and also a {stringi} wrapper to make the generated file much less icky.

I need to add in a test for invalid input or a codecov noop comment to sget the codecov % back to passing.

**Comment by hrbrmstr** [comment]
7408a20 shld say "fold" vs "find". I felt icky having something that cld not run on all platforms so rewrote `fold` in inline Rcpp in the script that generates the header.

File: tests/testthat/test-tz-lookup-edge.R [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+  "Antarctica/McMurdo", "America/Los_Angeles", "Australia/Darwin",
+  "Africa/Addis_Ababa", "Australia/Perth", "Australia/Perth", "Antarctica/McMurdo",
+  "America/New_York", "America/Belem", "Antarctica/Syowa", "Antarctica/Troll",
+  "Antarctica/McMurdo", "Antarctica/Vostok", "Asia/Yakutsk", "Asia/Vladivostok",
+  "Europe/Bucharest", "America/Iqaluit", "America/Cambridge_Bay",
+  "America/Godthab", "America/Argentina/Salta") -> val
+
+test_that("tz_lookup_coords edge cases work", {
+
+  for (idx in seq_along(lat)) {
```
Comment (ateucher) [inline tests/testthat/test-tz-lookup-edge.R:572]: Since this is vectorized, I'm not sure we need to loop through this? I think it could just be:

```r
expect_equal(
      tz_lookup_coords(lat, lng, "fast", FALSE),
      val
)
```

File: data-raw/generate-vars.R [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+library(stringi)
+library(Rcpp)
+
+tf <- tempfile(fileext = ".js")
+on.exit(unlink(tf), add=TRUE)
+
+# Grab the pre-computed string tree and mapped TZ list
+download.file(
+  "https://raw.githubusercontent.com/darkskyapp/tz-lookup/master/tz.js",
```
Comment (ateucher) [inline data-raw/generate-vars.R:13]: This is downloading the latest `tz.js`, but the "accurate" method is currently using the [`2018d` version from time-zone-boundary-builder](https://github.com/evansiroky/timezone-boundary-builder/releases/tag/2018d), so I think that is probably the source of some of the differences when comparing "fast" and "accurate".

I tried quickly with (what I think is) [the analogous `tz.js`](https://raw.githubusercontent.com/darkskyapp/tz-lookup/01136c2e0de64ea5380c1f71aedd5cec351ba348/tz.js), but it looks like the array is constructed slightly differently so the regex-wrangling didn't quite work... I'm happy to look at this further, as I think it would be good to use the same version in "fast" and "accurate"

File: tests/testthat/test-tz-lookup-edge.R (hunk 3 above)
Comment (ateucher) [reply tests/testthat/test-tz-lookup-edge.R:572]: And holy smokes is it ever fast!

**Comment by ateucher** [comment]
@hrbrmstr thanks again and sorry for the radio silence. I should be able to look at this more in depth next week, but I left a couple of inline comments - certainly not expecting you to do the work, but wanted to capture some thoughts, and at least contribute a little to your awesome work! I love the script to auto-generate `src/generated-vars.h`, especially the `fold` function 😄 .

Regarding perfect alignment between the `"fast"` and `"accurate"` methods, I'm also not too fussed about some being named, and some being GMT offsets... I think a lot will be resolved if we get them both using the latest version of the timezone boundaries, as I mentioned in my comment. I have started using the latest boundary file for the "accurate" method in the [update-2019a](https://github.com/ateucher/lutz/tree/update-2019a) branch, but ran into an issue with overlapping timezones which I haven't yet decided how to deal with (I don't think these are an issue (yet) in the darksky boundaries - see https://github.com/darkskyapp/tz-lookup/issues/34).

File: data-raw/generate-vars.R (hunk 4 above)
Comment (ateucher) [reply data-raw/generate-vars.R:13]: This was sorted out by moving both methods to use the latest (2019b) boundary in #2.

File: tests/testthat/test-tz-lookup-edge.R (hunk 3 above)
Comment (ateucher) [reply tests/testthat/test-tz-lookup-edge.R:572]: I'll leave this as is so when/if one fails we can easily pinpoint it
